File: backend/tour/elevation.js
/*
发送GET请求“https://maps.googleapis.com/maps/api/elevation/json?locations=latitude,longitude[|lati,longi|...]”；
每天2500次免费请求，按客户端与服务器端查询次数之和计算；
每次请求512个位置；
每秒50次请求，按客户端与服务器端查询次数之和计算。

估计——
单线程每秒请求一次，一次500点，2500秒（不到42分钟）完成一天的采集任务，共计125万点；
实际空间约东西8700米、南北9900米；
最大精度点容量约东西92394点、南北105004点。
逐行扫描法，从西往东、由北向南：
步长选择0.000001度时，每天可扫描13.5行，共需7778天多！
依此看来：要么将步长放大到0.001以上；要么增加有效线程。
*/

/*********************
****! Deprecated !****
*********************/

//速度调节因子
var factor = 3; //取值范围是[0,6]（值越大越快）——TODO:这种加速稀疏采样的方式对数据精度损伤太大导致位置偏离显著

//采点目标区域（矩形）
var top = parseFloat(34.543466.toFixed(factor));
var bottom = parseFloat(34.438462.toFixed(factor));
var right = parseFloat(110.115926.toFixed(factor));
var left = parseFloat(110.023532.toFixed(factor));
var step = 0.000001*(Math.pow(10,factor));
var deltay = top*(1000000/Math.pow(10,factor)) - bottom*(1000000/Math.pow(10,factor));
var deltax = right*(1000000/Math.pow(10,factor)) - left*(1000000/Math.pow(10,factor));

var url = "http://maps.googleapis.com/maps/api/elevation/json?";

var request = require("request"); //TODO: 从http到request还是没解决httproxy配置的问题（对node相关模块不熟悉吧）
request.defaults({"proxy":"http://127.0.0.1:8087"});

function fetch(query) {
    request(url+query, function (error, response, body) {
        console.log(error);
        console.log(response);
        console.log(body);
    });
}

//扫描
for (var y=0, cursory=top; y<deltay; y++, cursory=Math.round(((cursory+step)*Math.pow(10,factor)))/Math.pow(10,factor)) {
    for (var x=0, cursorx=left; x<deltax; x++, cursorx=Math.round(((cursorx+step)*Math.pow(10,factor)))/Math.pow(10,factor)) {
        query = "locations=" + cursory + "," + cursorx;
        fetch(query);
        break;
    }
    break;
}

File: backend/tour/elevation.py
#!/usr/bin/env python
#coding=utf-8

import requests
import json
import sqlite3
import time

K = 2000 # URL长度
L = 0.1 # 防止过快拒绝
M = 50 # 一次请求的点数
N = 3 # 累计错误容忍量

db = sqlite3.connect("./data.db")
cursor = db.cursor()
cursor.execute("CREATE TABLE MOUNTHUA(latitude TEXT, longitude TEXT, elevation TEXT);")
db.commit()

url = "http://maps.googleapis.com/maps/api/elevation/json"
proxies = {"http": "127.0.0.1:8087"}

north = 34543466
south = 34438462
east = 110115926
west = 110023532
step = 1
scale = 1000000.0 # 防止还原时丢失小数部分

def fetch():
    def dotgen():
        for latitude in xrange(south, north, step):
            for longitude in xrange(west, east, step):
                yield latitude/scale, longitude/scale
    limit = M
    tries = N
    positions = []
    with open(".\data.js", "w") as log:
        print >> log, time.ctime()
        for position in dotgen():
            if not tries:
                break
            positions.append("%s,%s" % position)
            if len(positions) % limit == 0:
                locations = "|".join(positions)
                length = len(locations)
                assert length < K
                params = {"locations": locations}
                response = requests.get(url, params=params, proxies=proxies)
                time.sleep(L)
                data = response.json()
                print >> log, "//", response.url
                print >> log, "// var data =", json.dumps(data)
                status = data.get("status")
                results = data.get("results")
                if status == "OK" and results:
                    for result in results:
                        location = result.get("location")
                        cursor.execute("INSERT INTO MOUNTHUA VALUES('%s', '%s', '%s')" % (location.get("lat"), location.get("lng"), result.get("elevation")))
                    db.commit()
                else:
                    print >> log, status
                    tries = tries - 1
                positions[:] = []
        print >> log, time.ctime()

try:
    fetch()
finally:
    cursor.close()
    db.close()